README.md
[![Build](https://github.com/sarmkadan/dotnet-saga-orchestrator/actions/workflows/build.yml/badge.svg)](https://github.com/sarmkadan/dotnet-saga-orchestrator/actions/workflows/build.yml)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![.NET](https://img.shields.io/badge/.NET-10.0-purple.svg)](https://dotnet.microsoft.com/)

# Distributed Saga Orchestrator for .NET Microservices

A production-ready distributed saga orchestrator for .NET microservices implementing the Saga pattern with compensating transactions, automatic retry logic, timeout handling, and comprehensive persistence. Built with .NET 10 and designed for enterprise-scale distributed systems.

## Table of Contents

- [Features](#features)
- [Architecture](#architecture)
- [Quick Start](#quick-start)
- [Installation](#installation)
- [Usage Examples](#usage-examples)
- [API Reference](#api-reference)
- [Configuration](#configuration)
- [CLI Reference](#cli-reference)
- [Compensation Strategies](#compensation-strategies)
- [Troubleshooting](#troubleshooting)
- [Testing](#testing)
- [Performance](#performance)
- [Related Projects](#related-projects)
- [Contributing](#contributing)
- [License](#license)

## Features

### Core Capabilities

- **Saga Orchestration**: Coordinate complex business transactions across multiple microservices with guaranteed consistency
- **Compensating Transactions**: Automatic rollback with five configurable compensation strategies for different failure scenarios
- **Retry Logic**: Exponential backoff with configurable retry policies per step
- **Timeout Handling**: Automatic detection and handling of step and saga-level timeouts
- **Persistence**: In-memory and file-based persistence options (database support planned for Phase 4)
- **Correlation IDs**: Distributed tracing through correlation IDs for request tracking across services
- **Event System**: Publisher-subscriber pattern for saga lifecycle events and webhooks
- **Rate Limiting**: Token bucket algorithm for protecting downstream services
- **Circuit Breaker**: Fault tolerance pattern preventing cascade failures
- **Metrics & Monitoring**: Real-time execution statistics and health checks

### Compensation Strategies

The orchestrator supports five compensation strategies:

1. **Reverse Order (LIFO)** - Default: Compensate in reverse order of completion
2. **Forward Order (FIFO)** - Compensate steps in execution order
3. **From Failure Point** - Compensate only failed step and subsequent steps
4. **Parallel** - Execute all compensations concurrently
5. **Manual** - External intervention for complex scenarios

### Saga Statuses

- **Pending** - Saga created but not initialized
- **Initialized** - Saga definition loaded, ready to start
- **Running** - Actively executing steps
- **Completed** - All steps succeeded
- **Failed** - One or more steps failed
- **Compensating** - Compensation in progress
- **Compensated** - Rollback completed
- **Aborted** - Manually aborted
- **TimedOut** - Exceeded overall timeout

## Architecture

```
┌────────────────────────────────────────────────────────┐
│          Presentation Layer (CLI Interface)             │
│         Commands: create, execute, status, etc.         │
└────────────────────────────────────────────────────────┘
                          ↓
┌────────────────────────────────────────────────────────┐
│  Application Layer (Services, Validators, Mappers)      │
│  - SagaOrchestrationService                             │
│  - SagaDefinitionService                                │
│  - CompensationService                                  │
└────────────────────────────────────────────────────────┘
                          ↓
┌────────────────────────────────────────────────────────┐
│  Infrastructure Layer (HTTP, Caching, Events, Workers)  │
│  - HttpClientFactory (with resilience)                  │
│  - EventBus (Pub/Sub)                                   │
│  - CacheService                                         │
│  - CircuitBreaker & RateLimiter                         │
└────────────────────────────────────────────────────────┘
                          ↓
┌────────────────────────────────────────────────────────┐
│       Core Layer (Domain Models, Extensions)            │
│  - Saga, SagaDefinition, SagaStep models                │
│  - 130+ utility extension methods                       │
└────────────────────────────────────────────────────────┘
                          ↓
┌────────────────────────────────────────────────────────┐
│      Data Layer (Repository Pattern)                    │
│  - InMemorySagaRepository                               │
│  - InMemorySagaStepRepository                           │
│  - InMemoryCompensationTransactionRepository            │
└────────────────────────────────────────────────────────┘
```

## Quick Start

### Prerequisites

- .NET 10 SDK ([download](https://dotnet.microsoft.com/download))
- Any text editor or IDE (VS Code, Visual Studio, Rider recommended)

### Installation

#### Method 1: Clone from Repository

```bash
git clone https://github.com/Sarmkadan/dotnet-saga-orchestrator.git
cd dotnet-saga-orchestrator
dotnet build
```

#### Method 2: Use as NuGet Package (Coming in Phase 4)

```bash
dotnet package add SagaOrchestrator
```

#### Method 3: Docker

```bash
docker build -t saga-orchestrator .
docker run --rm saga-orchestrator
```

### Basic Example

```csharp
using Microsoft.Extensions.DependencyInjection;
using SagaOrchestrator.Application.Services;
using SagaOrchestrator.Configuration;

// Setup dependency injection
var services = new ServiceCollection();
services.AddSagaOrchestrator();
var provider = services.BuildServiceProvider();

// Get services
var definitionService = provider.GetRequiredService<SagaDefinitionService>();
var orchestration = provider.GetRequiredService<SagaOrchestrationService>();

// Create saga definition
var definition = await definitionService.CreateDefinitionAsync(
    "Order Processing",
    "Process orders across microservices");

// Add steps
var step = new SagaStepDefinition(
    "Reserve Inventory",
    "inventory-service",
    "http://inventory/reserve",
    "http://inventory/release");
await definitionService.AddStepAsync(definition.Id, step);

// Execute saga
var saga = await orchestration.CreateSagaAsync(definition);
await orchestration.StartSagaAsync(saga.Id);

// Execute steps
for (int i = 0; i < saga.Steps.Count; i++)
{
    await orchestration.ExecuteNextStepAsync(saga.Id);
}
```

## Usage Examples

### Example 1: E-Commerce Order Processing

Complete order processing saga across inventory, payment, and shipping services:

```csharp
var definition = await definitionService.CreateDefinitionAsync(
    "E-Commerce Order Processing",
    "Process customer orders with inventory, payment, and shipping");

// Step 1: Reserve inventory
var reserveStep = new SagaStepDefinition(
    "Reserve Inventory",
    "inventory-service",
    "http://inventory:5001/api/reserve",
    "http://inventory:5001/api/release");
reserveStep.SetTimeout(30);
reserveStep.SetRetryPolicy(3, 1000);
await definitionService.AddStepAsync(definition.Id, reserveStep);

// Step 2: Charge payment
var paymentStep = new SagaStepDefinition(
    "Process Payment",
    "payment-service",
    "http://payment:5002/api/charge",
    "http://payment:5002/api/refund");
paymentStep.SetTimeout(30);
paymentStep.SetRetryPolicy(2, 2000);
await definitionService.AddStepAsync(definition.Id, paymentStep);

// Step 3: Create shipment
var shippingStep = new SagaStepDefinition(
    "Create Shipment",
    "shipping-service",
    "http://shipping:5003/api/create",
    "http://shipping:5003/api/cancel");
shippingStep.SetTimeout(60);
shippingStep.SetRetryPolicy(3, 1000);
await definitionService.AddStepAsync(definition.Id, shippingStep);

var saga = await orchestration.CreateSagaAsync(definition);
await orchestration.StartSagaAsync(saga.Id);
```

### Example 2: Financial Transfer Saga

Cross-account transfer with account validation and ledger updates:

```csharp
var definition = await definitionService.CreateDefinitionAsync(
    "Financial Transfer",
    "Transfer funds between accounts");

var validateStep = new SagaStepDefinition(
    "Validate Accounts",
    "account-service",
    "http://accounts/validate",
    "http://accounts/unlock");
await definitionService.AddStepAsync(definition.Id, validateStep);

var debitStep = new SagaStepDefinition(
    "Debit Source Account",
    "ledger-service",
    "http://ledger/debit",
    "http://ledger/credit");
debitStep.SetRetryPolicy(5, 500);
await definitionService.AddStepAsync(definition.Id, debitStep);

var creditStep = new SagaStepDefinition(
    "Credit Destination Account",
    "ledger-service",
    "http://ledger/credit",
    "http://ledger/debit");
creditStep.SetRetryPolicy(5, 500);
await definitionService.AddStepAsync(definition.Id, creditStep);

var saga = await orchestration.CreateSagaAsync(definition, maxRetries: 5);
await orchestration.StartSagaAsync(saga.Id);
```

### Example 3: Hotel Booking Saga

Multi-step booking process with hotel, flight, and car rental:

```csharp
var builder = new SagaStepBuilder()
    .WithName("Book Hotel")
    .WithServiceName("hotel-service")
    .WithExecutionUrl("http://hotel/book")
    .WithCompensationUrl("http://hotel/cancel")
    .WithTimeout(45)
    .WithRetries(3, 1000)
    .WithCircuitBreaker(5, 30);

var definition = await definitionService.CreateDefinitionAsync(
    "Travel Booking",
    "Book hotel, flight, and car rental");

await definitionService.AddStepAsync(definition.Id, builder.Build());

// Similar steps for flights and car rental...
```

### Example 4: Monitoring & Webhooks

Subscribe to saga events and send webhooks:

```csharp
var eventBus = provider.GetRequiredService<EventBus>();
var webhookHandler = provider.GetRequiredService<WebhookHandler>();

// Subscribe to saga completion
eventBus.Subscribe<SagaCompletedEvent>(async @event =>
{
    var webhook = new WebhookSubscription(
        "https://webhook.example.com/saga-completed",
        new[] { "saga.completed" });
    
    await webhookHandler.SendWebhookAsync(
        webhook,
        @event,
        retries: 3);
});

// Subscribe to failures
eventBus.Subscribe<SagaFailedEvent>(async @event =>
{
    // Send alert to monitoring system
    logger.LogError($"Saga {event.SagaId} failed: {event.FailureReason}");
});
```

### Example 5: Custom Retry Policy

Exponential backoff with jitter:

```csharp
var step = new SagaStepDefinition(
    "Call External API",
    "api-service",
    "http://api/process",
    "http://api/rollback");

step.RetryPolicy = new RetryPolicy
{
    MaxRetries = 5,
    InitialDelayMs = 100,
    BackoffMultiplier = 2.0,
    MaxDelayMs = 30000
};

await definitionService.AddStepAsync(definition.Id, step);
```

### Example 6: Compensation Strategy

Configure parallel compensation for independent steps:

```csharp
var saga = await orchestration.CreateSagaAsync(
    definition,
    compensationStrategy: CompensationStrategy.Parallel);

// If saga fails, all compensations execute concurrently
await orchestration.CompensateSagaAsync(saga.Id);
```

### Example 7: Health Monitoring

Real-time health checks and metrics:

```csharp
var healthService = provider.GetRequiredService<HealthCheckService>();
var metricsService = provider.GetRequiredService<MetricsService>();

var health = await healthService.GetHealthAsync();
Console.WriteLine($"Service Status: {health.ServiceStatus}");
Console.WriteLine($"Active Sagas: {health.ActiveSagaCount}");

var metrics = metricsService.GetMetrics();
Console.WriteLine($"Total Sagas: {metrics.TotalSagas}");
Console.WriteLine($"Success Rate: {metrics.SuccessRate:P2}");
Console.WriteLine($"P95 Duration: {metrics.P95DurationMs}ms");
```

### Example 8: Circuit Breaker Integration

Automatic fault tolerance:

```csharp
var circuitBreaker = provider.GetRequiredService<CircuitBreaker>();

try
{
    var result = await circuitBreaker.ExecuteAsync(
        "payment-service",
        async () =>
        {
            return await httpClient.PostAsync("http://payment/charge", content);
        });
}
catch (OpenCircuitException ex)
{
    logger.LogWarning($"Circuit breaker open for {ex.ServiceName}");
    // Fall back to alternative service
}
```

### Example 9: Rate Limiting

Prevent overwhelming downstream services:

```csharp
var rateLimiter = provider.GetRequiredService<RateLimiter>();

var result = await rateLimiter.AllowRequestAsync("payment-service", rps: 10);
if (!result.Allowed)
{
    throw new ServiceUnavailableException(
        $"Rate limit exceeded. Retry after {result.RetryAfterSeconds}s");
}
```

### Example 10: Caching Results

Cache step responses for performance:

```csharp
var cacheService = provider.GetRequiredService<CacheService>();

var cacheKey = CacheKeyBuilder.BuildSagaKey(sagaId);
var cached = await cacheService.GetAsync<Saga>(cacheKey);

if (cached != null)
{
    return cached;
}

var saga = await orchestration.GetSagaAsync(sagaId);
await cacheService.SetAsync(cacheKey, saga, expirationMinutes: 5);
return saga;
```

## API Reference

### SagaOrchestrationService

```csharp
public class SagaOrchestrationService
{
    // Create a new saga instance from definition
    Task<Saga> CreateSagaAsync(SagaDefinition definition, 
        int maxRetries = 3, int timeoutSeconds = 300);

    // Start saga execution
    Task<Saga> StartSagaAsync(string sagaId);

    // Execute the next step
    Task<SagaStep> ExecuteNextStepAsync(string sagaId);

    // Get saga by ID
    Task<Saga> GetSagaAsync(string sagaId);

    // List all sagas with optional filter
    Task<List<Saga>> ListSagasAsync(SagaStatus? status = null);

    // Initiate compensation
    Task<Saga> CompensateSagaAsync(string sagaId, 
        CompensationStrategy strategy = CompensationStrategy.ReverseOrder);

    // Abort saga execution
    Task<Saga> AbortSagaAsync(string sagaId, string reason = "");
}
```

### SagaDefinitionService

```csharp
public class SagaDefinitionService
{
    // Create a new saga definition
    Task<SagaDefinition> CreateDefinitionAsync(string name, 
        string description = "");

    // Get definition by ID
    Task<SagaDefinition> GetDefinitionAsync(string definitionId);

    // Add step to definition
    Task<SagaStepDefinition> AddStepAsync(string definitionId, 
        SagaStepDefinition step);

    // List all definitions
    Task<List<SagaDefinition>> ListDefinitionsAsync();

    // Validate definition
    ValidationResult ValidateDefinition(SagaDefinition definition);

    // Update definition
    Task<SagaDefinition> UpdateDefinitionAsync(string definitionId, 
        SagaDefinition definition);
}
```

### CompensationService

```csharp
public class CompensationService
{
    // Execute compensation for a saga
    Task<CompensationResult> CompensateSagaAsync(string sagaId, 
        CompensationStrategy strategy);

    // Get compensation status
    Task<CompensationTransaction> GetCompensationStatusAsync(string sagaId);

    // List all compensations
    Task<List<CompensationTransaction>> ListCompensationsAsync(
        CompensationStatus? status = null);
}
```

## Configuration

### Default Configuration

```csharp
public static class SagaConstants
{
    public const int DefaultSagaTimeoutSeconds = 300;      // 5 minutes
    public const int DefaultStepTimeoutSeconds = 30;       // 30 seconds
    public const int DefaultMaxRetries = 3;                // 3 attempts
    public const int DefaultRetryDelayMs = 1000;           // 1 second
    public const int DefaultCircuitBreakerFailureThreshold = 5;
    public const int DefaultCircuitBreakerTimeoutSeconds = 30;
    public const int DefaultRateLimitPerSecond = 100;
}
```

### Custom Configuration

```csharp
var services = new ServiceCollection();
var options = services.AddSagaOrchestrator()
    .WithDefaultStepTimeout(60)
    .WithDefaultMaxRetries(5)
    .WithCachingEnabled(true, ttlMinutes: 10)
    .WithWebhooksEnabled(true)
    .WithTimeoutWorker(true, checkIntervalSeconds: 10)
    .WithCompensationWorker(true, checkIntervalSeconds: 5)
    .WithExponentialBackoff(true, baseDelayMs: 100, maxDelayMs: 30000)
    .Build();
```

### Environment Variables

```bash
# Timeout configuration
SAGA_TIMEOUT_SECONDS=300
SAGA_STEP_TIMEOUT_SECONDS=30

# Retry configuration
SAGA_MAX_RETRIES=3
SAGA_RETRY_DELAY_MS=1000

# Feature flags
SAGA_ENABLE_CACHING=true
SAGA_ENABLE_WEBHOOKS=true
SAGA_ENABLE_TIMEOUT_WORKER=true

# Service configuration
SAGA_LOG_LEVEL=Information
SAGA_CORRELATION_ID_HEADER=X-Correlation-ID
```

## CLI Reference

The orchestrator provides a command-line interface for managing sagas.

### Commands

#### Create Saga

```bash
dotnet run -- create --definition "OrderProcessing" --timeout 300
```

Options:
- `--definition` (required): Saga definition name
- `--timeout` (optional): Saga timeout in seconds (default: 300)

#### Execute Step

```bash
dotnet run -- execute --saga <saga-id>
```

Options:
- `--saga` (required): Saga ID to execute next step for

#### Status

```bash
dotnet run -- status --saga <saga-id> --verbose
```

Options:
- `--saga` (required): Saga ID
- `--verbose` (optional): Show detailed step information

#### List Sagas

```bash
dotnet run -- list --status Running --limit 10
```

Options:
- `--status` (optional): Filter by status (Pending, Running, Completed, Failed, etc.)
- `--limit` (optional): Limit results (default: all)

#### Compensate

```bash
dotnet run -- compensate --saga <saga-id> --strategy ReverseOrder
```

Options:
- `--saga` (required): Saga ID
- `--strategy` (optional): Strategy (ReverseOrder, ForwardOrder, Parallel, Manual)

#### Help

```bash
dotnet run -- help
```

## Compensation Strategies

### Reverse Order (LIFO) - Default

Compensate in reverse order of completion:

```
Execution: Step1 → Step2 → Step3 ✓ ✓ ✗
Compensation: Step2 ← Step3 (skip failed step)
```

Use when: Step dependencies form a chain (e.g., reserve → charge → ship)

### Forward Order (FIFO)

Compensate in execution order:

```
Execution: Step1 → Step2 → Step3 ✓ ✓ ✗
Compensation: Step1 → Step2 → Step3
```

Use when: Compensation order doesn't matter or steps are independent

### From Failure Point

Only compensate from failure onward:

```
Execution: Step1 → Step2 → Step3 ✓ ✓ ✗
Compensation: (skip Step1, compensate Step2 and Step3)
```

Use when: Earlier steps are unaffected by later failures

### Parallel

Execute all compensations concurrently:

```
Execution: Step1 → Step2 → Step3 ✓ ✓ ✗
Compensation: Step1 ∥ Step2 ∥ Step3 (concurrent)
```

Use when: Compensations are independent and performance is critical

### Manual

Halt and require external intervention:

```
Execution: Step1 → Step2 → Step3 ✓ ✓ ✗
Compensation: [AWAITING_MANUAL_INTERVENTION]
```

Use when: Compensation requires human decision-making

## Troubleshooting

### Common Issues

#### Saga Timeout

**Problem**: Saga exceeds timeout without completing.

**Solution**:
```csharp
// Increase timeout
var saga = await orchestration.CreateSagaAsync(definition, timeoutSeconds: 600);

// Or increase step timeout
step.SetTimeout(60);
```

#### Step Failures

**Problem**: Specific step consistently fails.

**Solution**:
```csharp
// Increase retries and delay
step.SetRetryPolicy(maxRetries: 5, initialDelayMs: 2000);

// Or add circuit breaker
step.SetCircuitBreaker(failureThreshold: 5, timeoutSeconds: 30);
```

#### Memory Leaks

**Problem**: Memory usage grows over time.

**Solution**:
```csharp
// Configure cache cleanup
services.AddCaching(ttlMinutes: 5, maxEntriesInMemory: 1000);

// Enable event archival
var eventWorker = provider.GetRequiredService<EventProcessingWorker>();
await eventWorker.ArchiveOldEventsAsync(olderThanMinutes: 60);
```

#### Deadlocks

**Problem**: Saga hangs indefinitely.

**Solution**:
```csharp
// Check correlation ID tracking
var correlationId = context.RequestContext.CorrelationId;
logger.LogInformation($"Processing saga with correlation: {correlationId}");

// Use compensation to recovery
await orchestration.CompensateSagaAsync(sagaId);
```

### Debug Logging

Enable detailed logging:

```csharp
services.AddLogging(config =>
{
    config.AddConsole();
    config.SetMinimumLevel(LogLevel.Debug);
    config.AddFilter("SagaOrchestrator", LogLevel.Debug);
});
```

### Performance Tuning

```csharp
// Enable caching
services.AddCaching(ttlMinutes: 10, maxEntriesInMemory: 5000);

// Configure rate limiting
var rateLimiter = provider.GetRequiredService<RateLimiter>();
rateLimiter.SetLimit("payment-service", requestsPerSecond: 50);

// Tune background workers
var timeoutWorker = provider.GetRequiredService<SagaTimeoutWorker>();
await timeoutWorker.SetCheckIntervalAsync(checkIntervalSeconds: 15);
```

## Testing

The test suite covers saga lifecycle, retry policies, compensation flows, and infrastructure utilities.

### Running Tests

```bash
# Run all tests
dotnet test

# Run with coverage report
dotnet test --collect:"XPlat Code Coverage"

# Run a specific test class
dotnet test --filter "FullyQualifiedName~SagaLifecycleTests"
```

### Test Structure

| Suite | Description |
|---|---|
| `SagaLifecycleTests` | Saga creation, step execution, status transitions, and abort flows |
| `RetryPolicyTests` | Exponential backoff, max-retry limits, and jitter behaviour |
| `InfrastructureAndExtensionsTests` | Circuit breaker, rate limiter, cache, and utility extensions |

### Writing Tests

```csharp
[Fact]
public async Task SagaCompletesSuccessfully_WhenAllStepsSucceed()
{
    // Arrange
    var services = new ServiceCollection();
    services.AddSagaOrchestrator();
    var provider = services.BuildServiceProvider();

    var definitionService = provider.GetRequiredService<SagaDefinitionService>();
    var orchestration   = provider.GetRequiredService<SagaOrchestrationService>();

    var definition = await definitionService.CreateDefinitionAsync("Test Saga");
    await definitionService.AddStepAsync(definition.Id,
        new SagaStepDefinition("Step 1", "svc", "http://svc/do", "http://svc/undo"));

    // Act
    var saga = await orchestration.CreateSagaAsync(definition);
    await orchestration.StartSagaAsync(saga.Id);
    await orchestration.ExecuteNextStepAsync(saga.Id);

    // Assert
    var result = await orchestration.GetSagaAsync(saga.Id);
    Assert.Equal(SagaStatus.Completed, result.Status);
}
```

## Performance

Benchmarked on a single core (AMD EPYC 7763, .NET 10, in-memory repositories):

| Scenario | Throughput / Latency |
|---|---|
| Saga creation + start | ~14,000 ops/sec |
| Single-step execution (in-memory) | ~12,000 sagas/sec |
| Step scheduling overhead (median) | <0.5 ms |
| Parallel compensation (10 steps) | <5 ms end-to-end |
| Cache read (in-process) | <0.1 ms |
| Health-check endpoint | <2 ms P99 |
| Memory per saga instance (at rest) | ~1.8 KB |

### Scaling Notes

- Background workers (`SagaTimeoutWorker`, `CompensationWorker`) run on dedicated threads and do not contend with the execution hot path.
- Switching from in-memory to a persistent repository (Phase 4) will reduce write throughput by roughly 5–10× depending on the database; the orchestration logic overhead remains the same.
- Rate limiting and circuit breaker checks add ~0.05 ms per step; they can be disabled per-step if not needed.

## Related Projects

- [dotnet-event-bus](https://github.com/sarmkadan/dotnet-event-bus) - In-process and distributed event bus for .NET - pub/sub, request/reply, dead letter, polymorphic handlers
- [dotnet-distributed-lock](https://github.com/sarmkadan/dotnet-distributed-lock) - Distributed locking library for .NET - Redis, SQLite, PostgreSQL backends with fencing tokens and auto-renewal

### Integration Examples

**Publish saga lifecycle events via `dotnet-event-bus`**

```csharp
// Wire the saga event bus to the distributed bus so other services react to outcomes
services.AddSagaOrchestrator();
services.AddDistributedEventBus(); // from dotnet-event-bus

var sagaBus = provider.GetRequiredService<EventBus>();

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.2] - 2026-03-19

### Fixed
- Fix compensating transaction not executing when step fails with timeout
- Added regression test for the fix

## [2.0.0] - 2026-03-18

### Added
- Add distributed saga debugger with time-travel inspection
- Docker support with multi-stage builds
- Health check endpoints (/health, /health/ready)
- Integration test suite with xUnit
- Migration guide from v1.x

### Changed
- Upgraded to .NET 10.0
- Modern C# features (records, primary constructors)
- Improved API consistency

### Fixed
- Various edge cases found through testing

## [Unreleased]

### Planned
- Database persistence (SQL Server, PostgreSQL) via Entity Framework Core
- REST API endpoints (ASP.NET Core)
- Distributed tracing (OpenTelemetry)
- Message queue integration (RabbitMQ, Kafka)
- Web dashboard for saga monitoring

## [2.0.0] - 2026-03-01

### Changed
- Default container port changed from 80 to 8080 (non-root best practice)
- Docker runtime base image switched from `dotnet/runtime` to `dotnet/aspnet` for health check middleware and future REST API support
- Multi-stage Dockerfile rewritten with layer-cached restore for faster rebuilds
- Docker Compose file updated to Compose Specification (removed deprecated `version` key)
- HEALTHCHECK start period increased to 10s for cold-start reliability
- Container user hardened with restricted shell (`/sbin/nologin`)

### Added
- `docs/MIGRATION_v2.md` - migration guide covering all breaking changes from v1.x to v2.0

### Breaking
- Port 80 is no longer used; update all port mappings and health check URLs to 8080
- `ASPNETCORE_URLS` now defaults to `http://+:8080`

## [1.0.0] - 2025-09-22

### Added
- Five complete example projects: OrderProcessing, MoneyTransfer, TravelBooking, AdvancedRetries, CompensationStrategies, MetricsMonitoring
- Comprehensive documentation: getting-started guide, API reference, architecture overview, deployment guide, FAQ
- Docker support with Dockerfile and docker-compose.yml
- GitHub Actions CI/CD workflow with build, test, and NuGet publish pipelines
- CodeQL security scanning workflow
- Dependabot configuration for automated dependency updates
- Makefile for build automation
- NuGet packaging metadata with README embed
- `.editorconfig` for consistent code style

### Changed
- Expanded README with 10+ usage examples covering all major features
- Improved error messages throughout for better developer experience
- Enhanced XML documentation on all public APIs

### Fixed
- Concurrent access race condition in `CacheService` under high parallelism
- Edge case in `SagaTimeoutWorker` where a saga completing at the exact timeout boundary could be incorrectly marked as timed out
- `RetryPolicy` validation now rejects zero-delay configurations that could cause tight busy-loops

## [0.9.0] - 2025-08-25

### Added
- Full xUnit test suite: `SagaLifecycleTests`, `RetryPolicyTests`, `InfrastructureAndExtensionsTests`
- `FluentAssertions` and `Moq` dependencies in test project
- Test project wired into solution with correct `ProjectReference`
- `SagaVisualizationService` for ASCII state-machine rendering
- `VisualizationServiceExtensions` for DI registration

### Fixed
- `CompensationWorker` could skip a compensation record when the repository enumeration was modified concurrently
- `EventProcessingWorker` subscriber leak on repeated subscribe/unsubscribe cycles

## [0.8.0] - 2025-08-04

### Added
- CLI interface (`CliHandler`, `SagaCliCommand`) with commands: `create`, `execute`, `status`, `list`, `compensate`, `help`
- `OutputFormatter` supporting JSON, CSV, and table output modes
- `SagaMessageTemplates` for consistent human-readable status messages
- `RequestContext` for per-request correlation ID propagation

### Changed
- `SagaOrchestrationService` now surfaces structured `SagaCommandResult` on every mutation, simplifying CLI and caller code

## [0.7.0] - 2025-07-14

### Added
- `MetricsService` with real-time counters: total sagas, success rate, P50/P95/P99 duration percentiles, compensation rate
- `HealthCheckService` reporting active saga count, worker liveness, and memory pressure
- `SagaDefinitionValidator` with fluent validation rules (duplicate step names, missing URLs, timeout sanity checks)
- `ValidationExtensions` helpers used throughout the application layer

### Fixed
- `SagaIdGenerator` now uses `Guid.NewGuid()` rather than a monotonic counter, preventing ID collisions across restarts

## [0.6.0] - 2025-06-23

### Added
- `SagaTimeoutWorker` background service that polls for and expires overdue sagas
- `CompensationWorker` background service for automatic compensation retry on transient failures
- `EventProcessingWorker` background service for async event fan-out
- `ServiceConfiguration` and `InfrastructureConfiguration` DI extension methods
- `SagaOptions` strongly-typed configuration class bound to `appsettings` / environment variables

### Changed
- Background workers run on dedicated `IHostedService` threads; they no longer contend with the main orchestration hot path

## [0.5.0] - 2025-06-02

### Added
- `CircuitBreaker` with configurable failure threshold and half-open recovery window
- `RateLimiter` using token-bucket algorithm; per-service limits configurable at runtime
- `CacheService` with TTL-based expiry and `CacheKeyBuilder` helpers
- `SagaJsonSerializer` wrapping `System.Text.Json` with saga-specific converters
- `ExceptionMapper` translating domain exceptions to HTTP status codes

### Fixed
- `HttpClientFactory` retry handler was not propagating `CancellationToken` to inner handlers

## [0.4.0] - 2025-05-12

### Added
- HTTP client factory (`HttpClientFactory`) with built-in retry and timeout resilience

examples/OrderProcessing.cs
#nullable enable
// =============================================================================
// Author: Vladyslav Zaiets | https://sarmkadan.com
// CTO & Software Architect
// =============================================================================

using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using SagaOrchestrator.Application.Services;
using SagaOrchestrator.Configuration;
using SagaOrchestrator.Core.Domain.Models;

/// Complete order processing saga example
/// Demonstrates: order creation, inventory reservation, payment processing, and shipment
public class OrderProcessingExample
{
    public static async Task Main(string[] args)
    {
        var services = new ServiceCollection();
        services.AddLogging(config =>
        {
            config.AddConsole();
            config.SetMinimumLevel(LogLevel.Information);
        });
        services.AddSagaOrchestrator();

        var serviceProvider = services.BuildServiceProvider();
        var definitionService = serviceProvider.GetRequiredService<SagaDefinitionService>();
        var orchestrationService = serviceProvider.GetRequiredService<SagaOrchestrationService>();
        var logger = serviceProvider.GetRequiredService<ILogger<OrderProcessingExample>>();

        try
        {
            logger.LogInformation("=== E-Commerce Order Processing Saga ===\n");

            // Create saga definition
            var definition = await definitionService.CreateDefinitionAsync(
                "Order Processing",
                "Complete order processing across inventory, payment, and shipping");

            logger.LogInformation("✓ Created definition: {Name}\n", definition.Name);

            // Add step 1: Reserve inventory
            var reserveStep = new SagaStepDefinition(
                "Reserve Inventory",
                "inventory-service",
                "http://localhost:5001/api/inventory/reserve",
                "http://localhost:5001/api/inventory/release");
            reserveStep.SetTimeout(30);
            reserveStep.SetRetryPolicy(3, 1000);
            await definitionService.AddStepAsync(definition.Id, reserveStep).ConfigureAwait(false);
            logger.LogInformation("✓ Added step: Reserve Inventory");

            // Add step 2: Process payment
            var paymentStep = new SagaStepDefinition(
                "Process Payment",
                "payment-service",
                "http://localhost:5002/api/payments/charge",
                "http://localhost:5002/api/payments/refund");
            paymentStep.SetTimeout(30);
            paymentStep.SetRetryPolicy(2, 2000);
            await definitionService.AddStepAsync(definition.Id, paymentStep).ConfigureAwait(false);
            logger.LogInformation("✓ Added step: Process Payment");

            // Add step 3: Create shipment
            var shipmentStep = new SagaStepDefinition(
                "Create Shipment",
                "shipping-service",
                "http://localhost:5003/api/shipments/create",
                "http://localhost:5003/api/shipments/cancel");
            shipmentStep.SetTimeout(60);
            shipmentStep.SetRetryPolicy(3, 1000);
            await definitionService.AddStepAsync(definition.Id, shipmentStep).ConfigureAwait(false);
            logger.LogInformation("✓ Added step: Create Shipment\n");

            // Validate definition
            var validation = definitionService.ValidateDefinition(definition);
            if (!validation.IsValid)
            {
                logger.LogError("✗ Validation failed");
                return;
            }
            logger.LogInformation("✓ Definition validation passed\n");

            // Get updated definition with steps
            var retrievedDef = await definitionService.GetDefinitionAsync(definition.Id).ConfigureAwait(false);

            // Create saga instance
            var saga = await orchestrationService.CreateSagaAsync(
                retrievedDef,
                maxRetries: 3,
                timeoutSeconds: 300);
            logger.LogInformation("✓ Created saga: {Id}", saga.Id);
            logger.LogInformation("  Correlation ID: {CorrelationId}", saga.CorrelationId);
            logger.LogInformation("  Status: {Status}\n", saga.Status);

            // Start saga
            var startedSaga = await orchestrationService.StartSagaAsync(saga.Id).ConfigureAwait(false);
            logger.LogInformation($"✓ Saga started");
            logger.LogInformation("  Total steps: {Count}\n", startedSaga.Steps.Count);

            // Execute steps
            logger.LogInformation("Executing steps...");
            for (int i = 0; i < startedSaga.Steps.Count; i++)
            {
                var step = await orchestrationService.ExecuteNextStepAsync(saga.Id).ConfigureAwait(false);
                if (step != null)
                {
                    logger.LogInformation($"✓ Step {i + 1}: {step.Name} - {step.Status}");
                    if (step.CompletedAt.HasValue)
                    {
                        var duration = step.CompletedAt.Value - step.StartedAt;
                        logger.LogInformation($"  Duration: {duration?.TotalSeconds:F2}s");
                    }
                }
            }

            // Get final state
            var finalSaga = await orchestrationService.GetSagaAsync(saga.Id).ConfigureAwait(false);
            logger.LogInformation("\n=== Final Status: {Status} ===", finalSaga.Status);
            logger.LogInformation($"Completed steps: {finalSaga.Steps.Count(s => s.Status == SagaStepStatus.Completed)}/{finalSaga.Steps.Count}");

            if (finalSaga.Status == SagaStatus.Completed)
            {
                logger.LogInformation("✓ Order processed successfully!");
            }
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "An error occurred");
        }
    }
}
